File: src-python/sphere.py
from hitable import Hitable
from hitable import HitRecord
from vector import Vector

class Sphere(Hitable):
    def __init__(self, center = Vector(), radius = 0.0):
        self.center = center
        self.radius = float(radius)

    def hit(self, ray, tMin = 0.0, tMax = float("inf")):
        oc = ray.origin - self.center
        a = ray.direction.dot(ray.direction)
        b = oc.dot(ray.direction)
        c = oc.dot(oc) - (self.radius * self.radius)
        discrim = (b * b) - (a * c)

        if discrim > 0.0:
            temp = (-b - (b * b - a * c) ** 0.5) / a
            if tMin < temp < tMax:
                t = temp
                p = ray.point_at_parameter(t)
                normal = (p - self.center) / Vector(self.radius, self.radius, self.radius)
                return HitRecord(t, p, normal)

            temp = (-b + (b * b - a * c) ** 0.5) / a
            if tMin < temp < tMax:
                t = temp
                p = ray.point_at_parameter(t)
                normal = (p - self.center) / Vector(self.radius, self.radius, self.radius)
                return HitRecord(t, p, normal)
        return None

File: Sources/Sphere.swift
import Foundation
#if !os(Linux)
import simd
#endif

final class Sphere: Hitable {
  private let center: Vector
  private let radius: Double
  private let material: Material

  init(center: Vector = Vector(), radius: Double = 0.0, material: Material) {
    self.center = center
    self.radius = radius
    self.material = material
  }

  private func hitRecord(_ t: Double, ray: Ray) -> HitRecord {
    let p = ray.pointAtParameter(t)
    let normal = (p - center) / radius
    return HitRecord(t: t, p: p, normal: normal, material: material)
  }

  func hit(ray: Ray, tMin: Double = 0.0, tMax: Double = Double.infinity) -> HitRecord? {
    let oc = ray.origin - center
    let a = ray.direction.dotp(ray.direction)
    let b = oc.dotp(ray.direction)
    let c = oc.dotp(oc) - (radius * radius)
    let discrim = (b * b) - (a * c)

    if discrim > 0.0 {
      let sqr = sqrt(discrim)
      let tmpOne = (-b - sqr) / a
      if tmpOne < tMax && tmpOne > tMin {
        return hitRecord(tmpOne, ray: ray)
      }

      let tmpTwo = (-b + sqr) / a
      if tmpTwo < tMax && tmpTwo > tMin {
        return hitRecord(tmpTwo, ray: ray)
      }
    }
    return nil
  }
}
